### ISSUE TITLE
chezmoi and VS Code remote containers dotfile support

### ISSUE THREAD

**Issue by thernstig** [twpayne/chezmoi#760]
## Is your feature request related to a problem? Please describe.

I am using VS Code as my main editor, and a growing pattern is to use [VS Code Remote - Containers](https://code.visualstudio.com/docs/remote/containers) (devcontainer.json) to spin up a development environment. If you work over multiple projects, with various technology stacks, this is a great way to setup each projects own stack.

But since the development is then run inside a container, I am missing my personal dotfiles for an efficient workflow. This is a perfect example for when dotfiles really shines.

VS Code dev containers has support for this, https://code.visualstudio.com/docs/remote/containers#_personalizing-with-dotfile-repositories, but the problem is that it expects it to work with e.g. a `~dotfiles/install.sh` command that it launches after having spun up the container.

I could find no information anywhere on how to get this to work in a seamless way with chezmoi. One problem is that chezmoi requires a binary. One option might be to have an install.sh file that actually installs chezmoi and runs the proper commands to setup the environment, but I am unsure that is the best way.

## Describe the solution you'd like

I am not sure of the solution, but it would be nice with a FAQ entry describing how to best achieve this. It is possible that even some feature should be added into chezmoi to support this, but I'm unsure how.

## Describe alternatives you've considered

An alternative is to use some other project than chezmoi to handle the dotfile management. But since chezmoi does so many things good, it'd be a shame to have to resort to that.

**Comment by twpayne** [comment]
Thanks for the suggestion. I haven't used VS Code Remote Containers myself, but reading through the docs I think you can do the following. I haven't tried it though.

In your dotfiles repo, create the following:

`install.sh`
```sh
#!/bin/sh

curl -sfL https://git.io/chezmoi | sh
bin/chezmoi apply
```

`.chezmoiignore`
```
install.sh
```

Make `install.sh` executable (`chmod a+x install.sh`) and commit it and `.chezmoiignore` to your repo and push the changes.

Then, add the following to your `settings.json`:

```json
{
  "dotfiles.repository": "your-github-id/your-dotfiles-repo",
  "dotfiles.targetPath": "~/.local/share/chezmoi",
  "dotfiles.installCommand": "~/.local/share/chezmoi/install.sh"
}
```

I _think_ this will work. It's possible that VS Code won't create the nested target path. If that's the case, please report here and I'll suggest a workaround.

I'll add this to the how-to once we've worked out how to do it :)

**Comment by thernstig** [comment]
How would this work with that basically every devcontainer could possibly use any type of linux distro or architecture?

What I mean is, one would most likely want to have a "bare minimum" happening when one does `bin/chezmoi apply`. How would one achieve this with `.chezmoi.toml.tmpl` to take this into account when installing? Before even knowing what system you will intall all your dotfiles into?

**Comment by twpayne** [comment]
> How would this work with that basically every devcontainer could possibly use any type of linux distro or architecture?

`curl -sfL https://git.io/chezmoi | sh` inspects the operating system and architecture of the system it's running on and downloads the appropriate prebuilt binary into `bin/`.

chezmoi is written in Go which produces standalone statically-linked binaries with almost no dependencies, which run on almost every Linux distribution.

> How would one achieve this with `.chezmoi.toml.tmpl` to take this into account when installing?

Ah, I see. The `install.sh` script above doesn't create your config file because it doesn't use chezmoi init. The following should work:

`install.sh`
```sh
#!/bin/sh

curl -sfL https://git.io/chezmoi | sh

mkdir -p ~/.config/chezmoi
bin/chezmoi execute-template --init --output ~/.config/chezmoi/chezmoi.toml < ~/.local/share/chezmoi/.chezmoi.toml.tmpl

bin/chezmoi apply
```

In the call to `chezmoi execute-template` you might also want to use the `--promptString` flag, see [the reference manual](https://github.com/twpayne/chezmoi/blob/master/docs/REFERENCE.md#execute-template-templates) for an explanation.

**Comment by thernstig** [comment]
Thanks! As I am new to chezmoi I'll need some time to grok this and learn a bit more before trying, but thanks for the pointers. https://www.chezmoi.io/docs/how-to/ and "_Create a config file on a new machine automatically_" had some good info regarding `--promptString`, but that example did not show the `--output` flag like you did above. But no worries I'll figure it out 😃 I am _guessing_ it has to do that I do not want the prompt during the init, but rather that it generates it beforehand.

Off-topic: https://www.chezmoi.io/docs/install/ says it installs to `./bin` and not `bin/` but I cannot confirm where that script installs since I used the `.deb` package. Could it be more clear where this  `.bin` is located? I could not find any info about it.

**Comment by twpayne** [comment]
Another option to consider is that `chezmoi archive` creates a tar archive of your dotfiles that you can untar anywhere. I don't think it's exactly what you want here, but it can be useful if (for example) your dotfiles contain secrets and you don't want to install your password manager in your target container.

> Off-topic: [chezmoi.io/docs/install](https://www.chezmoi.io/docs/install/) says it installs to `./bin` and not `bin/` but I cannot confirm where that script installs since I used the `.deb` package. Could it be more clear where this `bin/` is located? I could not find any info about it.

`./bin` and `bin/` are the same `bin` subdirectory written in two different ways. The `./` prefix is a way of expressing that the path is relative to the current subdirectory, and the `/` suffix is a way of expressing that it's a directory, not a file. The docs can be improved here too, thanks for spotting this :)

**Comment by thernstig** [comment]
I am aware of the differences in style for subdirectories, my mistake being unclear that I was really after the complete path for where the binaries were installed.

Personally, I am using `~.local/bin` for all user bins which is part of the systemd file-hierarchy spec (an extension to the xdg-user-dirs spec). Anyway I apologize again for going off-topic even in my own thread, will stop doing that...

**Comment by twpayne** [comment]
For info you can specify the installation path, for example:

```sh
curl -sfL https://git.io/chezmoi | sh -s -- -b ~/.local/bin
```

**Comment by felipecassiors** [comment]
The #818 is kind of a dupe of this. Didn't see this one when I created that other. I have my working example with `VS Code - dev containers` there in case you find it useful.

**Comment by twpayne** [comment]
#834 (merged) and #848 (will be merged very soon) should fix this.
